src/plugins/midi-piano/piano-component.js

```diff
@@ -0,0 +1,115 @@
+import PropTypes from 'prop-types';
+import { Piano } from 'react-piano';
+import React, { useEffect, useState, useRef } from 'react';
+
+export default function PianoComponent(props) {
+
+  const defaultKeyWidth = 29.2;
+  const renderCount = useRef(0);
+  const renderPiano = useRef(false);
+  const containerDiv = useRef(null);
+  const containerWidth = useRef(null);
+  const pianoWrapperDiv = useRef(null);
+  const pianoWrapperWidth = useRef(null);
+  const { noteRange, playNote, stopNote, keys } = props;
+  const [pianoWrapperDimensions, setPianoWrapperDimensions] = useState({});
+
+  function getNumberOfKeysRendered() {
+    return noteRange.last - noteRange.first + 1;
+  }
+
+  const numberOfKeysRendered = getNumberOfKeysRendered();
+
+  const getPianoWrapperDimensions = clientWidth => {
+    let width;
+    let height;
+    if (numberOfKeysRendered > 39) {
+      width = clientWidth;
+      const keyWidth = clientWidth / numberOfKeysRendered;
+      height = keyWidth * 5.5;
+      pianoWrapperWidth.current = width;
+      return { width: `${width}px`, height: `${height}px` };
+    }
+    const neededWidth = numberOfKeysRendered * defaultKeyWidth;
+    if (neededWidth > clientWidth) {
+      width = clientWidth;
+      const keyWidth = clientWidth / numberOfKeysRendered;
+      height = keyWidth * 5.5;
+    } else {
+      height = 160;
+      width = numberOfKeysRendered * defaultKeyWidth;
+    }
+    pianoWrapperWidth.current = width;
+    return { width: `${width}px`, height: `${height}px` };
+  };
+
+  const handleWindowResize = () => {
+    const neededWidth = numberOfKeysRendered * defaultKeyWidth;
+    if (containerDiv.current.clientWidth === containerWidth.current) {
+      return;
+    }
+    if (containerDiv.current.clientWidth < containerWidth.current && containerDiv.current.clientWidth > pianoWrapperWidth.current) {
+      return;
+    }
+    if (containerDiv.current.clientWidth > containerWidth.current && neededWidth < containerDiv.current.clientWidth) {
+      return;
+    }
+    const obj = getPianoWrapperDimensions(containerDiv.current.clientWidth);
+    setPianoWrapperDimensions(obj);
+  };
+
+  useEffect(() => {
+    renderCount.current += 1;
+  });
+
+  useEffect(() => {
+    if (containerDiv.current.clientWidth === containerWidth.current) {
+      return;
+    }
+    containerWidth.current = containerDiv.current.clientWidth;
+    const obj = getPianoWrapperDimensions(containerDiv.current.clientWidth);
+    renderPiano.current = true;
+    setPianoWrapperDimensions(obj);
+  });
+
+  useEffect(() => {
+    window.addEventListener('resize', handleWindowResize);
+    return function cleanUp() {
+      window.removeEventListener('resize', handleWindowResize);
+    };
+  }, []);
+
+  useEffect(() => {
+    if (renderCount.current !== 2) {
+      return;
+    }
+    const keyElems = document.querySelectorAll('.ReactPiano__Key');
```

Comment:
We should not use classes top get content like this. Does the piano player not have a way of exposing which keys it contains through props?
I wanted to check the documentation of `react-piano` but I'm not sure which package this is (please see https://github.com/educandu/educandu/pull/649/files#r989082791)

Reply:
It's this package: [https://www.npmjs.com/package/react-piano]

The keys to be shown have to be passed in via the prop `noteRange={ first: midiValueStart, last: midiValueEnd  }` so you would always know which keys are set to be shown. The problem is:

When playing back a midi file or when playing with a midi input device, I'd like to visualize the keys played by changing the color of the plugin-piano keys. The react-piano accepts the native prop `activeKeys` where you can pass in an array with the midi values of the keys to be changed in color. Since this is changing the state, the piano would rerender after every note which is played or released. It does kind of work, but because of the vast amount of rerenders in short time, the playback of the midi file or the playback of the keys you play gets cracks and missing notes and everything. This is why I changed the key visualization to DOM manipulation to get around these rerenders. Now the playback works as it's supposed to.
Is this an allowed thing to do? I can imagine that working with react you should prefer managing the state over of manipulating DOM elements, but this doesn't seem to be the right solution for playing music notes.

Reply:
There are basically two ways to work with such components: either they're 100% managed by React - or - they are 0% managed by React (meaning it's pure Vanilla DOM manipulation with only a React wrapper `div`). This mix doesn't really work well, because you never know when React will override your manual manipulations of the tree that is under React's control. I would first go the 100% pure React way. If the re-rendering gets too inefficient, there's always the possibility to either fix `react-piano` itself, or, if not possible/feasible, create our own piano component (it's just a bunch of `div` elements anyway, not that much magic).

Reply:
That makes sense! So for going 100% React, it depends on whether the package can be fixed so that the rerenders don't obstruct the audio playback. If this is possible, I must say that it surpasses what I can do. Although it doesn't mean too much, I can't imagine how that could be accomplished. Playing a chord consisting of only three notes on a midi device can be enough to cause problems in the audio playback.
As much as I would like to work with React, I would enjoy making the piano with Vanilla JS from ground up just the same. Maybe it would even be a good way to make sure that the functionality can be extended in the future without the risk of new issues coming up.
What would you suggest as the next steps?

Reply:
My 2 cents would be to try and address the other comments and low hanging fruit to clear up the code and get it to work so we could try it out too (see https://github.com/educandu/educandu/pull/649#issuecomment-1271538587). Then we could think a bit and talk how to best proceed. Or? What do you think, @ahelmberger ?

Reply:
I agree. Let's make it run first (using the current approach), so we can get a better picture of what the plugin will do. Then we can decide how to replace the piano (I think a vanilla custom-written version might be the most lightweight while still performant solution).

Reply:
Ok I'll go on and keep you informed on the progress.